beacon-chain/types/crystallized_state.go

```diff
@@ -241,33 +278,57 @@ func (c *CrystallizedState) CalculateNewCrystallizedState(aState *ActiveState, s
 
 	// Construct new crystallized state for cycle transition.
 	newCrystallizedState := NewCrystallizedState(&pb.CrystallizedState{
+		ShardAndCommitteesForSlots: c.data.ShardAndCommitteesForSlots,
+		CrosslinkingStartShard:     c.data.CrosslinkingStartShard,
+		DynastyStart:               c.data.DynastyStart,
+		CurrentDynasty:             c.data.CurrentDynasty,
+		DynastySeed:                c.data.DynastySeed,
 		Validators:                 rewardedValidators,
 		LastStateRecalc:            lastStateRecalc + params.CycleLength,
-		ShardAndCommitteesForSlots: newShardAndCommitteesForSlots,
 		LastJustifiedSlot:          justifiedSlot,
 		JustifiedStreak:            justifiedStreak,
 		LastFinalizedSlot:          finalizedSlot,
-		CrosslinkingStartShard:     0, // TODO(#494): Stub. Need to see where this epoch left off.
 		CrosslinkRecords:           newCrossLinkRecords,
-		DynastySeedLastReset:       c.data.DynastySeedLastReset, // TODO(#267): Stub. Need to implement dynasty transition.
 		TotalDeposits:              nextCycleBalance,
 	})
 
 	return newCrystallizedState, nil
 }
 
-func (c *CrystallizedState) calculateNewShardAndCommitteesForSlots() ([]*pb.ShardAndCommitteeArray, error) {
-	newCommittees, err := casper.ShuffleValidatorsToCommittees(
-		c.DynastySeed(),
-		c.Validators(),
-		c.CurrentDynasty(),
-		c.CrosslinkingStartShard(),
+// NewDynastyRecalculations recomputes the validator set. This method is called during a dynasty transition.
+func (c *CrystallizedState) NewDynastyRecalculations(seed [32]byte, slotNumber uint64) (*CrystallizedState, error) {
+	lastSlot := len(c.data.ShardAndCommitteesForSlots) - 1
+	lastCommitteeFromLastSlot := len(c.ShardAndCommitteesForSlots()[lastSlot].ArrayShardAndCommittee) - 1
+	crosslinkLastShard := c.ShardAndCommitteesForSlots()[lastSlot].ArrayShardAndCommittee[lastCommitteeFromLastSlot].ShardId % params.ShardCount
+	crosslinkNextShard := crosslinkLastShard + 1
```

Comment:
Your modular arithmetic can be off here.

Say `ShardCount == 10` and the `crosslinkLastShard` is 9, the above code would then set `crosslinkNextShard` equal to 10 instead of 0.

Code should be:
```go
crosslinkLastShard := c.ShardAndCommitteesForSlots()[lastSlot].ArrayShardAndCommittee[lastCommitteeFromLastSlot].ShardId
crosslinkNextShard := (crosslinkLastShard + 1) % params.ShardCount
```